### ISSUE TITLE
Dirty property with custom setter is not persisted in 6.1.x

### ISSUE THREAD

**Issue by dpcasady** [grails/grails-data-mapping#917]
In GORM 6.1.x, properties with custom/explicit setters are not marked as dirty and do not cause updates when the session is flushed. Properties that don't have custom setters work as expected.

```groovy
class Book {
    String title
    String author

    void setAuthor(String author) {
        this.author = author
    }
}
```

```groovy
Book b = new Book(author: 'My Author')

Book.withTransaction {
    b.save()
    b.author = 'Updated Author'
}

// in 6.0.x the db shows the author as 'Updated Author', in 6.1.x it's 'My Author'
```

In the example above, when the session is flushed, the `author` property is not seen as dirty and no update takes place. With the same code, but using 6.0.x, both the save and update occur.

Relevant Hibernate logging in 6.0.x where it works:
```
TRACE - o.h.e.i.AbstractSaveEventListener        : Transient instance of: dirty.checking.example.Book
TRACE - o.h.e.i.DefaultSaveOrUpdateEventListener : Saving transient instance
TRACE - o.h.e.i.AbstractSaveEventListener        : Saving [dirty.checking.example.Book#<null>]
TRACE - o.h.e.i.AbstractFlushingEventListener    : Flushing session
DEBUG - o.h.e.i.AbstractFlushingEventListener    : Processing flush-time cascades
DEBUG - o.h.e.i.AbstractFlushingEventListener    : Dirty checking collections
TRACE - o.h.e.i.AbstractFlushingEventListener    : Flushing entities and processing referenced collections
TRACE - o.h.e.i.DefaultFlushEntityEventListener  : Found dirty properties [[dirty.checking.example.Book#1]] : [Ljava.lang.String;@e83d546
TRACE - o.h.e.i.DefaultFlushEntityEventListener  : Updating entity: [dirty.checking.example.Book#1]
TRACE - o.h.e.i.AbstractFlushingEventListener    : Processing unreferenced collections
TRACE - o.h.e.i.AbstractFlushingEventListener    : Scheduling collection removes/(re)creates/updates
DEBUG - o.h.e.i.AbstractFlushingEventListener    : Flushed: 0 insertions, 1 updates, 0 deletions to 1 objects
DEBUG - o.h.e.i.AbstractFlushingEventListener    : Flushed: 0 (re)creations, 0 updates, 0 removals to 0 collections
TRACE - o.h.e.i.AbstractFlushingEventListener    : Executing flush
TRACE - o.h.e.i.AbstractFlushingEventListener    : Post flush
```

6.1.x where it doesn't work:
```
TRACE - o.h.e.i.AbstractSaveEventListener        : Transient instance of: dirty.checking.example.Book
TRACE - o.h.e.i.DefaultSaveOrUpdateEventListener : Saving transient instance
TRACE - o.h.e.i.AbstractSaveEventListener        : Saving [dirty.checking.example.Book#<null>]
TRACE - o.h.e.i.AbstractFlushingEventListener    : Flushing session
DEBUG - o.h.e.i.AbstractFlushingEventListener    : Processing flush-time cascades
DEBUG - o.h.e.i.AbstractFlushingEventListener    : Dirty checking collections
TRACE - o.h.e.i.AbstractFlushingEventListener    : Flushing entities and processing referenced collections
TRACE - o.h.e.i.AbstractFlushingEventListener    : Processing unreferenced collections
TRACE - o.h.e.i.AbstractFlushingEventListener    : Scheduling collection removes/(re)creates/updates
DEBUG - o.h.e.i.AbstractFlushingEventListener    : Flushed: 0 insertions, 0 updates, 0 deletions to 1 objects
DEBUG - o.h.e.i.AbstractFlushingEventListener    : Flushed: 0 (re)creations, 0 updates, 0 removals to 0 collections
TRACE - o.h.e.i.AbstractFlushingEventListener    : Executing flush
TRACE - o.h.e.i.AbstractFlushingEventListener    : Post flush
```

[Example grails 3.2.8/GORM 6.1.1 app](https://github.com/dpcasady/dirty-checking-example) that demonstrates the issue. Run the integration tests and you should see the failure.

**Comment by graemerocher** [comment]
Can be worked around by calling `markDirty('author', author)` from within the setter

**Comment by dpcasady** [comment]
Great, thanks!

**Comment by dpcasady** [comment]
Just out of curiosity, if adding a property to a domain via an AST transformation, it's just a matter of adding `markDirty('propName', propName)` in the setter? Unless it's added prior to canonicalization it won't be picked up by the DirtyCheckTransformation, right?

**Comment by demon101** [comment]
@graemerocher  could you also port this fix to 6.0.x branch, please

**Comment by graemerocher** [comment]
Right
